subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -127,12 +129,73 @@ def _get_plan_for_product_id(self, product_id: str) -> Plan:
+        # Latest transaction could be "CANCELLED", as it might be an upgrade.
+        # We assume that the user has a single subscription active for this app on the Apple platform.
         return SubscriptionPayment.objects.filter(
             provider_codename=self.codename,
             metadata__original_transaction_id=original_transaction_id,
         ).order_by('subscription_end').last()
 
+    def _get_active_transaction(self, transaction_id: str, original_transaction_id: str) -> SubscriptionPayment:
+        """
+        Note: this should be a simple `get`, however the database could contain duplicates.
+        Once all duplicates are defeated, replace this with `get`.
+        """
+        obtained_entries = SubscriptionPayment.objects.filter(
+            provider_codename=self.codename,
+            provider_transaction_id=transaction_id,
+            metadata__original_transaction_id=original_transaction_id,
+            status=SubscriptionPayment.Status.COMPLETED,
+        )
+        if len(obtained_entries) == 0:
+            raise SubscriptionPayment.DoesNotExist()
+
+        if len(obtained_entries) > 1:
+            logger.warning('Multiple active transactions found for transaction id "%s". '
+                           'Consider cleaning it up.', transaction_id)
+
+        return obtained_entries.first()
+
+    def _get_or_create_payment(self,
+                               transaction_id: str,
+                               original_transaction_id: str,
+                               user: User,
+                               plan: Plan,
+                               start: datetime.datetime,
+                               end: datetime.datetime,
+                               subscription: Optional[Subscription] = None) -> SubscriptionPayment:
+        try:
+            payment, was_created = SubscriptionPayment.objects.get_or_create(
+                provider_codename=self.codename,
+                provider_transaction_id=transaction_id,
+                defaults={
+                    'status': SubscriptionPayment.Status.COMPLETED,
+                    # In-app purchase doesn't report the money.
+                    # We mark it as None to indicate we don't know how much did it cost.
+                    'amount': None,
+                    'user': user,
+                    'plan': plan,
+                    'subscription': subscription,
+                    'subscription_start': start,
+                    'subscription_end': end,
+                }
+            )
+            if was_created:
+                payment.subscription.auto_prolong = False
+                # Note: initial transaction is the one that has the same original transaction id and transaction id.
+                payment.meta = AppleInAppMetadata(original_transaction_id=original_transaction_id)
+                payment.save()
+        except SubscriptionPayment.MultipleObjectsReturned:
```

Comment:
Critical! `payment.save()` does not save `payment.subscription`, so `auto_prolong` won't be saved as `False`